sql/catalyst/src/main/scala/org/apache/spark/sql/Row.scala

```diff
@@ -314,7 +314,7 @@ trait Row extends Serializable {
    *
    * @throws ClassCastException when data type does not match.
    */
-  def getSeq[T](i: Int): Seq[T] = getAs[Seq[T]](i)
+  def getSeq[T](i: Int): scala.collection.Seq[T] = getAs[scala.collection.Seq[T]](i)
```

Comment:
Can you explain this fix a bit more? I don't doubt it, but do we need to promise a `scala.collection.Seq`?

Reply:
The return type or convert type?There are some things that need to be discussed. There are 2 major changes related to scala 2.13：

- Seq` alias representative `scala.collection.immutable.Seq`,  not  `scala.collection.Seq`

- `WrappedArray.make` method produce `mutable.ArraySeq` instance, not `WrappedArray`

Explicit use cast type as getAs[scala.collection.Seq[T]](i) because `ArrayTpye` all mapped to `scala.collection.Seq` in `RowEncoder`, otherwise, the cast will fail because of the actual data type is `mutable.ArraySeq`, the affected cases are `RowEncoder should support primitive arrays` and `RowEncoder should support primitive arrays` in `RowEncoderSuite`.

Explicit appoint return type to `scala.collection.Seq` because we promise a `scala.collection.Seq` in Scala 2.12 and use `Seq` alias will trigger a `toSeq` operation, but it can be removed if we are sure  covert to `scala.collection.immutable.Seq` doesn't affect the caller's perception.

Reply:
On the other hand, we can use `scala.collection.immutable.ArraySeq.unsafeWrapArray` to instead of `WrappedArray.make` in Scala 2.13 to produce a `immutable.ArraySeq`,then we remove the changes here.
But `immutable.ArraySeq.unsafeWrapArray` not support in Scala 2.12, this will causes a code branch to distinguish between Scala versions in `RowEncoder.deserializerFor` and we need use reflect api to get `immutable.ArraySeq` class as follow:

```
if (isScala213) {
        StaticInvoke(
          Utils.classForName("scala.collection.immutable.ArraySeq"),
          ObjectType(classOf[Seq[_]]),
          "unsafeWrapArray",
          arrayData :: Nil,
          returnNullable = false)
} else {
....
}
```

Reply:
On the other hand, we can use `immutable.ArraySeq.unsafeWrapArray` to instead of `mutable.WrappedArray.make` in Scala 2.13 to produce a `immutable.ArraySeq`,then  maybe we can remove the changes here, but `immutable.ArraySeq.unsafeWrapArray` not support in Scala 2.12

Reply:
Yeah, there's a parallel problem though. If someone writes `val foo: Seq[String] = getSeq[String](0)`, then in Scala 2.13, they are expecting an immutable Seq. However this method returns `scala.collection.Seq` now, breaking the code. Because all Scala apps would move to Seq meaning immutable Seq, I think we actually want to follow that as it helps compatibility. 

It means Scala 2.13 builds aren't necessarily source or binary compatible with 2.12, but, that has never been true for any Scala app.

I think it's OK to encode arrays as (immutable) Seq even if it means wrapping them, calling .toSeq or whatever, but that is perhaps the most controversial part of a change like this.

Yes, the goal is to avoid parallel source trees if we can avoid it. It's easier to manage and maybe allows fewer places where behavior might vary. That means it won't be possible to use Scala 2.13-only classes, ideally.

Reply:
Address [5944512](https://github.com/apache/spark/pull/29434/commits/594451208426eb3b30bcd9a2b4cce74f25621e8e) revert to use `Seq` alias as `Row.getSeq` method return type.

The internal `ArrayType -> scala.collection.Seq` encode mode remain unchanged